scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala

```diff
@@ -39,55 +40,102 @@ class RemoveUnused(config: RemoveUnusedConfig)
     }
 
   override def fix(implicit doc: SemanticDocument): Patch = {
-    val isUnusedTerm = mutable.Set.empty[Position]
-    val isUnusedImport = mutable.Set.empty[Position]
+    val unusedTermsPosition = mutable.Set.empty[Position]
+    val unusedImportsPosition = mutable.Set.empty[Position]
 
     doc.diagnostics.foreach { diagnostic =>
       if (config.imports && diagnostic.message == "Unused import") {
-        isUnusedImport += diagnostic.position
+        unusedImportsPosition += diagnostic.position
       } else if (config.privates &&
         diagnostic.message.startsWith("private") &&
         diagnostic.message.endsWith("is never used")) {
-        isUnusedTerm += diagnostic.position
+        unusedTermsPosition += diagnostic.position
       } else if (config.locals &&
         diagnostic.message.startsWith("local") &&
         diagnostic.message.endsWith("is never used")) {
-        isUnusedTerm += diagnostic.position
+        unusedTermsPosition += diagnostic.position
       }
     }
 
-    if (isUnusedImport.isEmpty && isUnusedTerm.isEmpty) {
+    if (unusedImportsPosition.isEmpty && unusedTermsPosition.isEmpty) {
       // Optimization: don't traverse if there are no diagnostics to act on.
       Patch.empty
     } else {
-      doc.tree.collect {
-        case i: Importee if isUnusedImport(importPosition(i)) =>
-          i match {
-            case Importee.Rename(_, to) =>
+      fix(doc.tree, unusedTermsPosition.toSet, unusedImportsPosition.toSet)
+    }
+  }
+
```

Comment:
Algorithm:
- traverses the tree using depth-first traversal
- for each enclosing scope (source, package, object, class, def, block, ...)
  - any used import wildcards AND unused import renames are tracked and passed down to nested scopes following them
  - any other unused import gets removed
  - at the end of each scope
    - any unused import rename (from outer our current scope) whose package is the same as any wildcard import in scope (outer or current) gets unimported
    - any remaining unused import rename belonging exclusively to the current scope (not the outer) gets removed